File: packages/workspace_flutter/ios/Classes/Internal/WorkspaceProtocol.swift
import Foundation

struct IOSVaultEnvelope: Equatable {
  let bookmark: Data
  let generation: UInt64
  let rootID: String
  let rootDigest: Data
}

/// Protocol-v1 validation shared by the channel handler and XCTest target.
enum WorkspaceProtocol {
  static let version = 1
  static let maxWorkspaceBytes = 512
  static let maxOperationIDBytes = 128
  static let maxStableIDBytes = 512
  static let maxEnvelopeBytes = 1_048_576
  static let maxRemainingMillis: Int64 = 86_400_000

  /// A platform-tagged binary vault credential. It binds the opaque bookmark
  /// to the native root generation and opaque root ID; Dart never decodes it.
  static func iOSEnvelope(
    bookmark: Data,
    generation: UInt64,
    rootID: String,
    rootDigest: Data
  ) -> Data? {
    guard !bookmark.isEmpty,
          let rootIDData = opaqueIDData(rootID),
          rootDigest.count == rootDigestBytes,
          bookmark.count <= maxEnvelopeBytes - envelopeHeaderSize - rootIDData.count else {
      return nil
    }
    var envelope = envelopeMagic
    envelope.append(iOSPlatform)
    var encodedGeneration = generation.bigEndian
    withUnsafeBytes(of: &encodedGeneration) { envelope.append(contentsOf: $0) }
    var rootLength = UInt16(rootIDData.count).bigEndian
    withUnsafeBytes(of: &rootLength) { envelope.append(contentsOf: $0) }
    envelope.append(rootIDData)
    envelope.append(rootDigest)
    var bookmarkLength = UInt32(bookmark.count).bigEndian
    withUnsafeBytes(of: &bookmarkLength) { envelope.append(contentsOf: $0) }
    envelope.append(bookmark)
    return envelope
  }

  static func iOSVaultEnvelope(fromEnvelope envelope: Data) -> IOSVaultEnvelope? {
    guard envelope.count > envelopeHeaderSize,
          envelope.prefix(envelopeMagic.count) == envelopeMagic,
          envelope[envelopeMagic.count] == iOSPlatform else { return nil }
    var index = envelopeMagic.count + 1
    guard let generation = readUInt64(envelope, index: &index),
          let rootLength = readUInt16(envelope, index: &index),
          rootLength > 0,
          index + Int(rootLength) + rootDigestBytes + 4 < envelope.count else { return nil }
    let rootData = envelope[index..<(index + Int(rootLength))]
    index += Int(rootLength)
    guard let rootID = String(data: rootData, encoding: .utf8), opaqueIDData(rootID) != nil else {
      return nil
    }
    let digest = Data(envelope[index..<(index + rootDigestBytes)])
    index += rootDigestBytes
    guard let bookmarkLength = readUInt32(envelope, index: &index),
          bookmarkLength > 0,
          Int(bookmarkLength) == envelope.count - index else { return nil }
    return IOSVaultEnvelope(
      bookmark: Data(envelope[index...]), generation: generation, rootID: rootID, rootDigest: digest
    )
  }

  static func hasVersion(_ values: [String: Any]) -> Bool {
    values["protocolVersion"] as? Int == version
  }

  static func hasExactKeys(_ values: [String: Any], _ expected: Set<String>) -> Bool {
    Set(values.keys) == expected
  }

  static func workspaceID(_ value: Any?) -> String? {
    guard let value = value as? String,
          !value.isEmpty,
          !value.contains("\0"),
          value.lengthOfBytes(using: .utf8) <= maxWorkspaceBytes else { return nil }
    return value
  }

  static func operationID(_ value: Any?) -> String? {
    guard let value = value as? String,
          !value.isEmpty,
          value.utf8.count <= maxOperationIDBytes,
          value.unicodeScalars.allSatisfy(isOpaqueCharacter) else { return nil }
    return value
  }

  static func stableID(_ value: Any?) -> String? {
    guard let value = value as? String,
          !value.isEmpty,
          value.utf8.count <= maxStableIDBytes,
          value.unicodeScalars.allSatisfy(isOpaqueCharacter) else { return nil }
    return value
  }

  static func boundedInt(_ value: Any?, minimum: Int, maximum: Int) -> Int? {
    guard let exact = exactInteger(value), exact >= Int64(minimum), exact <= Int64(maximum) else {
      return nil
    }
    return Int(exact)
  }

  static func nonNegativeInt64(_ value: Any?) -> Int64? {
    guard let exact = exactInteger(value), exact >= 0 else { return nil }
    return exact
  }

  static func remainingMillis(_ value: Any?) -> Int64? {
    guard let raw = exactInteger(value), raw >= 1, raw <= maxRemainingMillis else { return nil }
    return raw
  }

  private static func isOpaqueCharacter(_ scalar: UnicodeScalar) -> Bool {
    switch scalar.value {
    case 45, 48...57, 65...90, 95, 97...122: return true
    default: return false
    }
  }

  private static let envelopeMagic = Data([0x57, 0x53, 0x45, 0x02])
  private static let iOSPlatform: UInt8 = 2
  private static let envelopeHeaderSize = 4 + 1 + 8 + 2 + 32 + 4
  private static let rootDigestBytes = 32

  private static func opaqueIDData(_ value: String) -> Data? {
    guard !value.isEmpty,
          value.utf8.count <= maxStableIDBytes,
          value.unicodeScalars.allSatisfy(isOpaqueCharacter) else { return nil }
    return Data(value.utf8)
  }

  private static func readUInt16(_ data: Data, index: inout Int) -> UInt16? {
    guard index + 2 <= data.count else { return nil }
    let value = data[index..<(index + 2)].reduce(UInt16(0)) { ($0 << 8) | UInt16($1) }
    index += 2
    return value
  }

  private static func readUInt32(_ data: Data, index: inout Int) -> UInt32? {
    guard index + 4 <= data.count else { return nil }
    let value = data[index..<(index + 4)].reduce(UInt32(0)) { ($0 << 8) | UInt32($1) }
    index += 4
    return value
  }

  private static func readUInt64(_ data: Data, index: inout Int) -> UInt64? {
    guard index + 8 <= data.count else { return nil }
    let value = data[index..<(index + 8)].reduce(UInt64(0)) { ($0 << 8) | UInt64($1) }
    index += 8
    return value
  }

  private static func exactInteger(_ value: Any?) -> Int64? {
    guard let number = value as? NSNumber else { return nil }
    let raw = number.int64Value
    // NSNumber can carry a floating-point value from malformed channel input.
    return number.doubleValue == Double(raw) ? raw : nil
  }
}

File: packages/workspace_flutter/android/src/main/kotlin/com/mattsp1290/workspace_flutter/internal/WorkspaceProtocol.kt
package com.mattsp1290.workspace_flutter.internal

import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets

internal data class AndroidVaultEnvelope(
  val treeUri: String,
  val generation: Long,
  val rootId: String,
  val rootDigest: ByteArray,
)

/** Protocol-v1 validation shared by the channel handler and JVM tests. */
internal object WorkspaceProtocol {
  const val version = 1
  const val maxWorkspaceBytes = 512
  const val maxOperationIdBytes = 128
  const val maxStableIdBytes = 512
  const val maxEnvelopeBytes = 1_048_576
  const val maxRemainingMillis = 86_400_000L

  fun workspaceId(value: Any?): String? = (value as? String)
    ?.takeIf { it.isNotEmpty() && it.toByteArray(Charsets.UTF_8).size <= maxWorkspaceBytes && !it.contains('\u0000') }

  fun operationId(value: Any?): String? = (value as? String)
    ?.takeIf { it.isNotEmpty() && it.toByteArray(Charsets.UTF_8).size <= maxOperationIdBytes && it.all(::isOpaqueCharacter) }

  fun stableId(value: Any?): String? = (value as? String)
    ?.takeIf { it.isNotEmpty() && it.toByteArray(Charsets.UTF_8).size <= maxStableIdBytes && it.all(::isOpaqueCharacter) }

  fun envelope(value: Any?): ByteArray? = (value as? ByteArray)
    ?.takeIf { it.isNotEmpty() && it.size <= maxEnvelopeBytes }

  /**
   * The vault credential is binary and platform tagged. Its URI payload remains
   * opaque to Dart; accepting bare UTF-8 here would make a tree URI a public
   * channel convention again.
   */
  fun androidEnvelope(
    uri: String,
    generation: Long,
    rootId: String,
    rootDigest: ByteArray,
  ): ByteArray? {
    val payload = uri.toByteArray(StandardCharsets.UTF_8)
    val rootIdBytes = rootId.toByteArray(StandardCharsets.UTF_8)
    if (payload.isEmpty() || generation <= 0 || stableId(rootId) == null || rootDigest.size != rootDigestBytes ||
      payload.size > maxEnvelopeBytes - envelopeHeaderSize - rootIdBytes.size
    ) return null
    return ByteBuffer.allocate(envelopeHeaderSize + rootIdBytes.size + payload.size)
      .put(envelopeMagic)
      .put(androidPlatform)
      .putLong(generation)
      .putShort(rootIdBytes.size.toShort())
      .put(rootIdBytes)
      .put(rootDigest)
      .putInt(payload.size)
      .put(payload)
      .array()
  }

  fun androidVaultEnvelope(value: Any?): AndroidVaultEnvelope? {
    val bytes = envelope(value) ?: return null
    if (bytes.size <= envelopeHeaderSize || !bytes.copyOfRange(0, envelopeMagic.size).contentEquals(envelopeMagic)) {
      return null
    }
    val buffer = ByteBuffer.wrap(bytes)
    buffer.position(envelopeMagic.size)
    if (buffer.get() != androidPlatform) return null
    val generation = buffer.long
    val rootLength = buffer.short.toInt() and 0xffff
    if (generation <= 0 || rootLength <= 0 || rootLength > buffer.remaining() - rootDigestBytes - 4) return null
    val rootIdBytes = ByteArray(rootLength)
    buffer.get(rootIdBytes)
    val rootId = runCatching { StandardCharsets.UTF_8.newDecoder().decode(ByteBuffer.wrap(rootIdBytes)).toString() }.getOrNull()
      ?: return null
    if (stableId(rootId) == null) return null
    val digest = ByteArray(rootDigestBytes)
    buffer.get(digest)
    val length = buffer.int
    if (length <= 0 || length != buffer.remaining()) return null
    val payload = ByteArray(length)
    buffer.get(payload)
    val uri = runCatching {
      StandardCharsets.UTF_8.newDecoder().decode(ByteBuffer.wrap(payload)).toString()
    }.getOrNull() ?: return null
    return AndroidVaultEnvelope(uri, generation, rootId, digest)
  }

  fun boundedInt(value: Any?, minimum: Int, maximum: Int): Int? {
    val number = exactLong(value) ?: return null
    return number.takeIf { it in minimum.toLong()..maximum.toLong() }?.toInt()
  }

  fun nonNegativeLong(value: Any?): Long? = exactLong(value)?.takeIf { it >= 0 }

  fun remainingMillis(value: Any?): Long? {
    val asLong = exactLong(value) ?: return null
    return asLong.takeIf { it in 1..maxRemainingMillis }
  }

  fun hasVersion(arguments: Map<*, *>): Boolean = arguments["protocolVersion"] == version

  fun hasExactKeys(arguments: Map<*, *>, expected: Set<String>): Boolean =
    arguments.keys.all { it is String } && arguments.keys == expected

  private fun isOpaqueCharacter(value: Char): Boolean =
    value in 'A'..'Z' || value in 'a'..'z' || value in '0'..'9' || value == '-' || value == '_'

  private val envelopeMagic = byteArrayOf('W'.code.toByte(), 'S'.code.toByte(), 'E'.code.toByte(), 2)
  private const val androidPlatform: Byte = 1
  private const val rootDigestBytes = 32
  private const val envelopeHeaderSize = 4 + 1 + 8 + 2 + rootDigestBytes + 4

  private fun exactLong(value: Any?): Long? {
    val number = value as? Number ?: return null
    val asLong = number.toLong()
    // StandardMessageCodec can represent several Number types. Do not silently
    // truncate fractional values or accept a wrapping conversion.
    return asLong.takeIf { number.toDouble() == asLong.toDouble() }
  }
}
